// asset-rack/compiled/rack.js
// Generated by CoffeeScript 1.4.0
(function() {
  var BufferStream, ClientRack, ConfigRack, EventEmitter, async, extend, fs, pathutil, pkgcloud, _ref,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  async = require('async');

  pkgcloud = require('pkgcloud');

  fs = require('fs');

  pathutil = require('path');

  _ref = require('./util'), BufferStream = _ref.BufferStream, extend = _ref.extend;

  ClientRack = require('./.').ClientRack;

  EventEmitter = require('events').EventEmitter;

  exports.Rack = (function(_super) {

    __extends(Rack, _super);

    function Rack(assets, options) {
      var asset, _i, _len,
        _this = this;
      Rack.__super__.constructor.call(this);
      if (options == null) {
        options = {};
      }
      this.maxAge = options.maxAge;
      this.allowNoHashCache = options.allowNoHashCache;
      this.on('complete', function() {
        return _this.completed = true;
      });
      this.on('newListener', function(event, listener) {
        if (event === 'complete' && _this.completed === true) {
          return listener();
        }
      });
      this.on('error', function(error) {
        if (_this.listeners('error').length === 1) {
          throw error;
        }
      });
      for (_i = 0, _len = assets.length; _i < _len; _i++) {
        asset = assets[_i];
        asset.rack = this;
      }
      this.assets = [];
      async.forEachSeries(assets, function(asset, next) {
        asset.on('error', function(error) {
          return next(error);
        });
        asset.on('complete', function() {
          if (asset.contents != null) {
            _this.assets.push(asset);
          }
          if (asset.assets != null) {
            _this.assets = _this.assets.concat(asset.assets);
          }
          return next();
        });
        asset.rack = _this;
        return asset.emit('start');
      }, function(error) {
        if (error != null) {
          return _this.emit('error', error);
        }
        return _this.emit('complete');
      });
    }

    Rack.prototype.createClientRack = function() {
      var clientRack;
      clientRack = new ClientRack;
      clientRack.rack = this;
      clientRack.emit('start');
      return clientRack;
    };

    Rack.prototype.addClientRack = function(next) {
      var clientRack,
        _this = this;
      clientRack = this.createClientRack();
      return clientRack.on('complete', function() {
        _this.assets.push(clientRack);
        return next();
      });
    };

    Rack.prototype.handle = function(request, response, next) {
      var handle,
        _this = this;
      response.locals({
        assets: this
      });
      handle = function() {
        var asset, check, _i, _len, _ref1;
        _ref1 = _this.assets;
        for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
          asset = _ref1[_i];
          check = asset.checkUrl(request.url);
          if (check) {
            return asset.respond(request, response);
          }
        }
        return next();
      };
      if (this.completed) {
        return handle();
      } else {
        return this.on('complete', handle);
      }
    };

    Rack.prototype.writeConfigFile = function(filename) {
      var asset, config, _i, _len, _ref1;
      config = {};
      _ref1 = this.assets;
      for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
        asset = _ref1[_i];
        config[asset.url] = asset.specificUrl;
      }
      return fs.writeFileSync(filename, JSON.stringify(config));
    };

    Rack.prototype.deploy = function(options, next) {
      var deploy,
        _this = this;
      options.keyId = options.accessKey;
      options.key = options.secretKey;
      deploy = function() {
        var assets, client;
        client = pkgcloud.storage.createClient(options);
        assets = _this.assets;
        if (options.provider === 'rackspace') {
          assets = _this.assets.concat(_this.assets[0]);
        }
        return async.forEachSeries(assets, function(asset, next) {
          var clientOptions, headers, key, stream, url, value, _ref1;
          console.log(asset.url);
          stream = null;
          if (asset.gzip) {
            stream = new BufferStream(asset.gzipContents);
          } else {
            stream = new BufferStream(asset.contents);
          }
          url = asset.specificUrl.slice(1, asset.specificUrl.length);
          headers = {};
          _ref1 = asset.headers;
          for (key in _ref1) {
            value = _ref1[key];
            headers[key] = value;
          }
          if (options.provider === 'amazon') {
            headers['x-amz-acl'] = 'public-read';
          }
          clientOptions = {
            container: options.container,
            remote: url,
            headers: headers,
            stream: stream
          };
          return client.upload(clientOptions, function(error) {
            if (error != null) {
              return next(error);
            }
            return next();
          });
        }, function(error) {
          if (error != null) {
            if (next != null) {
              return next(error);
            }
            throw error;
          }
          if (options.configFile != null) {
            _this.writeConfigFile(options.configFile);
          }
          if (next != null) {
            return next();
          }
        });
      };
      if (this.completed) {
        return deploy();
      } else {
        return this.on('complete', deploy);
      }
    };

    Rack.prototype.tag = function(url) {
      var asset, _i, _len, _ref1;
      _ref1 = this.assets;
      for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
        asset = _ref1[_i];
        if (asset.url === url) {
          return asset.tag();
        }
      }
      throw new Error("No asset found for url: " + url);
    };

    Rack.prototype.url = function(url) {
      var asset, _i, _len, _ref1;
      _ref1 = this.assets;
      for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
        asset = _ref1[_i];
        if (url === asset.url) {
          return asset.specificUrl;
        }
      }
    };

    Rack.extend = extend;

    return Rack;

  })(EventEmitter);

  ConfigRack = (function() {

    function ConfigRack(options) {
      if (options.configFile == null) {
        throw new Error('options.configFile is required');
      }
      if (options.hostname == null) {
        throw new Error('options.hostname is required');
      }
      this.assetMap = require(options.configFile);
      this.hostname = options.hostname;
    }

    ConfigRack.prototype.handle = function(request, response, next) {
      var specificUrl, url, _ref1;
      response.locals({
        assets: this
      });
      _ref1 = this.assetMap;
      for (url in _ref1) {
        specificUrl = _ref1[url];
        if (request.url === url || request.url === specificUrl) {
          return response.redirect("//" + this.hostname + specificUrl);
        }
      }
      return next();
    };

    ConfigRack.prototype.tag = function(url) {
      var tag;
      switch (pathutil.extname(url)) {
        case '.js':
          tag = "\n<script type=\"text/javascript\" ";
          return tag += "src=\"//" + this.hostname + this.assetMap[url] + "\"></script>";
        case '.css':
          return "\n<link rel=\"stylesheet\" href=\"//" + this.hostname + this.assetMap[url] + "\">";
      }
    };

    ConfigRack.prototype.url = function(url) {
      return "//" + this.hostname + this.assetMap[url];
    };

    return ConfigRack;

  })();

  exports.fromConfigFile = function(options) {
    return new ConfigRack(options);
  };

}).call(this);
